README.md
# Transbank

[![Latest Version on Packagist][ico-version]][link-packagist]
[![Total Downloads][ico-downloads]][link-downloads]
[![Build Status][ico-travis]][link-travis]
[![StyleCI][ico-styleci]][link-styleci]

This is where your description should go. Take a look at [contributing.md](contributing.md) to see a to do list.

## Installation

Via Composer

``` bash
$ composer require innovaweb/transbank
```

## Usage

## Change log

Please see the [changelog](changelog.md) for more information on what has changed recently.

## Testing

``` bash
$ composer test
```

## Contributing

Please see [contributing.md](contributing.md) for details and a todolist.

## Security

If you discover any security related issues, please email aisla@innovaweb.cl instead of using the issue tracker.

## Credits

- [Alejandro Isla][link-author]
- [All Contributors][link-contributors]

## License

MIT. Please see the [license file](license.md) for more information.

[ico-version]: https://img.shields.io/packagist/v/innovaweb/transbank.svg?style=flat-square
[ico-downloads]: https://img.shields.io/packagist/dt/innovaweb/transbank.svg?style=flat-square
[ico-travis]: https://img.shields.io/travis/innovaweb/transbank/master.svg?style=flat-square
[ico-styleci]: https://styleci.io/repos/12345678/shield

[link-packagist]: https://packagist.org/packages/innovaweb/transbank
[link-downloads]: https://packagist.org/packages/innovaweb/transbank
[link-travis]: https://travis-ci.org/innovaweb/transbank
[link-styleci]: https://styleci.io/repos/12345678
[link-author]: https://github.com/innovaweb
[link-contributors]: ../../contributors


changelog.md
# Changelog

All notable changes to `Transbank` will be documented in this file.
## v2.10.0 - 2023-04-10

### Added

- OneClick Delete Inscription function

## v2.9.0 - 2022-04-29

### Added

- Laravel illuminate/support 9
- HelperRedirect function to return data array
- Refund Response Codes
- Response Commerce Code Response Codes

### Updated

- VCI Response Codes

### Fixed

- Source Code
- GuzzleException

---

## v2.8.0 - 2022-04-29

### Added

- Laravel illuminate/support 9
- HelperRedirect function to return data array
- Refund Response Codes
- Response Commerce Code Response Codes

### Updated

- VCI Response Codes

### Fixed

- Source Code
- GuzzleException

---

## v2.7.0 - 2022-04-29

### Added

- Laravel illuminate/support 9
- HelperRedirect function to return data array
- Refund Response Codes
- Response Commerce Code Response Codes

### Updated

- VCI Response Codes

### Fixed

- Source Code

---

## v2.6.0 - 2022-04-28

### Added

- Laravel illuminate/support 9
- HelperRedirect function to return data array

### Updated

- Transbank SDK ^2.0

---

## v2.5.0 - 2022-01-03

### Fixed

- get status transaction by token webpay plus

---

## v2.4.0 - 2022-01-03

### Update

- Fixed get status transaction by token webpay plus

---

## v2.3.0 - 2021-10-04

### Update

- Added code commerce params on oneclick authorize

---

## v2.2.0 - 2021-07-26

### Update

- Added installments number to pay in one click mall

---

## v2.1.0 - 2021-07-20

### Update

- Update versión composer

---

## v2.0.1 - 2021-07-20

### Added

- Update commerce code one click mall develop mode

---

## v2.0.0 - 2021-07-20

### Added

- Update transbank php sdk from 1.x to 2.0.6

---

## v1.2.0 - 2021-06-22

### Added

- Oneclick Mall API REST Integration

---

## v1.1.1 - 2021-04-21

### Fixed

- Se cambia a "transbank/transbank-sdk": "~1.0"

---

## v1.1.0 - 2021-04-21

### Fixed

- Se revierte la compatibilidad a transbank-php-sdk 1.10.1

---

## v1.0.0

### Update

- Update and fix transbank sdk ~2.0
- Update readme

---

## v0.3-beta

### Fix

- Production update sdk

---

## v0.2-beta

### Fix

- Production create transaction

---

## v0.1-beta

### Added

- Webpay Plus API REST integration




src/OneClickMall.php
<?php


namespace Innovaweb\Transbank;

use Exception;
use GuzzleHttp\Exception\GuzzleException;
use Innovaweb\Transbank\Helpers\HelperRedirect;
use Transbank\Webpay\Oneclick as OneClick;
use Transbank\Webpay\Oneclick\MallInscription as MallInscription;
use Transbank\Webpay\Oneclick\MallTransaction as MallTransaction;

class OneClickMall
{
    /** @const INTEGRATION Development environment */
    const INTEGRATION = 'integration';

    /** @const PRODUCTION Production environment */
    const PRODUCTION = 'production';

    public $token;
    public $url;
    public $commerce_code = 597055555543;

    public function __construct(
        $webpay_plus_commerce_code = '',
        $webpay_plus_api_key = '',
        $environment = self::INTEGRATION
    ) {
        if ($environment === self::PRODUCTION and !empty($webpay_plus_commerce_code) and !empty($webpay_plus_api_key)) {
            OneClick::configureForProduction($webpay_plus_commerce_code, $webpay_plus_api_key);
            $this->commerce_code = $webpay_plus_commerce_code;
        } else {
            OneClick::configureForTesting();
        }
    }

    /**
     * createInscription
     *
     */
    public function createInscription($username, $email, $url_return)
    {

        try {

            $response = (new MallInscription)->start($username, $email, $url_return);
            $this->token = $response->getToken();
            $this->url = $response->getUrlWebpay();

            return [
                'status' => 'success',
                'response' => $response
            ];
        } catch (Exception $exception) {
            return [
                'status' => 'error',
                'exception' => $exception->getMessage(),
            ];
        } catch (GuzzleException $exception) {
            return [
                'status' => 'error',
                'exception' => $exception->getMessage(),
            ];
        }
    }

    /**
     * finishInscription
     *
     */
    public function finishInscription($tbk_token)
    {
        try {
            if (!$tbk_token) {
                $tbk_token = $_POST['TBK_TOKEN'];
            }
            $response = (new MallInscription)->finish($tbk_token);

            if ((int)$response->getResponseCode() === 0) {
                return [
                    'status' => 'success',
                    'response' => $response
                ];
            } else {
                return [
                    'status' => 'error',
                    'exception' => $response,
                ];
            }
        } catch (Exception $exception) {
            return [
                'status' => 'error',
                'exception' => $exception->getMessage(),
            ];
        }
    }

    /**
     * authorize
     */
    public function authorize($username, $tbkUser, $order_id, $details)
    {

        try {
            $response = (new MallTransaction)->authorize($username, $tbkUser, $order_id, $details);
            return [
                'status' => 'success',
                'response' => $response
            ];
        } catch (Exception $exception) {
            return [
                'status' => 'error',
                'exception' => $exception->getMessage(),
            ];
        }
    }

    /**
     * getStatus
     *
     */
    public function getStatus($buyOrder)
    {
        try {
            $response = (new MallTransaction)->status($buyOrder);
            return [
                'status' => 'success',
                'response' => $response
            ];
        } catch (Exception $exception) {
            return [
                'status' => 'error',
                'exception' => $exception->getMessage(),
            ];
        }
    }


    /**
     * deleteInscription
     *
     */
    public function deleteInscription($tbkUser, $username)
    {
        try {
            $response = (new MallInscription)->delete($tbkUser, $username);

            return [
                'status' => 'success',
                'response' => $response
            ];
        } catch (Exception $exception) {
            return [
                'status' => 'error',
                'exception' => $exception->getMessage(),
            ];
        } catch (GuzzleException $exception) {
            return [
                'status' => 'error',
                'exception' => $exception->getMessage(),
            ];
        }
    }

    public function redirectHTML($url = '', $token = '')
    {
        $url = empty($url) ? $this->url : $url;
        $token = empty($token) ? $this->token : $token;
        return HelperRedirect::redirectHTML($url, $token, 'oneclick');
    }
}


src/Helpers/HelperRedirect.php
<?php

namespace Innovaweb\Transbank\Helpers;

class HelperRedirect
{

    public static function redirectHTML($url, $token = '', $type = 'webpay')
    {
        if ($type == 'webpay') {
            return self::htmlWrapWebPay($url, $token);
        } else {
            return self::htmlWrapOneClick($url, $token);
        }
    }

    private static function htmlWrapWebPay($url, $token)
    {
        $formName = 'webpay-redirect-form-' . uniqid();

        return
            '<html lang="es">
            <head>
                <title>Conectando con Transbank</title>
            </head>
            <body>
                <form action="' . $url . '" id="' . $formName . '" method="POST">
                    <input type="hidden" name="token_ws" value="' . $token . '" />
                </form>
                <script>document.getElementById("' . $formName . '").submit();</script>
            </body>
        </html>';
    }

    private static function htmlWrapOneClick($url, $token)
    {
        $formName = 'oneclick-redirect-form-' . uniqid();

        return
            '<html lang="es">
            <head>
                <title>Conectando con Transbank</title>
            </head>
            <body>
                <form action="' . $url . '" id="' . $formName . '" method="POST">
                    <input type="hidden" name="TBK_TOKEN" value="' . $token . '" />
                </form>
                <script>document.getElementById("' . $formName . '").submit();</script>
            </body>
        </html>';
    }

    private static function dataArray($url, $token)
    {
        return [
            'url' => $url,
            'token' => $token
        ];
    }

}


src/TransbankServiceProvider.php
<?php

namespace Innovaweb\Transbank;

use Illuminate\Support\ServiceProvider;

class TransbankServiceProvider extends ServiceProvider
{
    /**
     * Perform post-registration booting of services.
     *
     * @return void
     */
    public function boot(): void
    {
        // $this->loadTranslationsFrom(__DIR__.'/../resources/lang', 'innovaweb');
        // $this->loadViewsFrom(__DIR__.'/../resources/views', 'innovaweb');
        // $this->loadMigrationsFrom(__DIR__.'/../database/migrations');
        // $this->loadRoutesFrom(__DIR__.'/routes.php');

        // Publishing is only necessary when using the CLI.
        if ($this->app->runningInConsole()) {
            $this->bootForConsole();
        }
    }

    /**
     * Register any package services.
     *
     * @return void
     */
    public function register(): void
    {
        $this->mergeConfigFrom(__DIR__.'/../config/transbank.php', 'transbank');

        // Register the service the package provides.
        $this->app->singleton('transbank', function ($app) {
            return new Transbank;
        });
    }

    /**
     * Get the services provided by the provider.
     *
     * @return array
     */
    public function provides()
    {
        return ['transbank'];
    }

    /**
     * Console-specific booting.
     *
     * @return void
     */
    protected function bootForConsole(): void
    {
        // Publishing the configuration file.
        $this->publishes([
            __DIR__.'/../config/transbank.php' => config_path('transbank.php'),
        ], 'transbank.config');

        // Publishing the views.
        /*$this->publishes([
            __DIR__.'/../resources/views' => base_path('resources/views/vendor/innovaweb'),
        ], 'transbank.views');*/

        // Publishing assets.
        /*$this->publishes([
            __DIR__.'/../resources/assets' => public_path('vendor/innovaweb'),
        ], 'transbank.views');*/

        // Publishing the translation files.
        /*$this->publishes([
            __DIR__.'/../resources/lang' => resource_path('lang/vendor/innovaweb'),
        ], 'transbank.views');*/

        // Registering package commands.
        // $this->commands([]);
    }
}


src/Helpers/HelperTransbankResponseCode.php
<?php


namespace Innovaweb\Transbank\Helpers;


class HelperTransbankResponseCode
{

    /**
     * VCI
     *
     * @param string $code vic.
     * @return string retorna el significado del código, Este campo es información adicional suplementaria al responseCode pero el comercio no debe validar este campo. Porque constantemente se agregan nuevos mecanismos de autenticación que se traducen en nuevos valores para este campo que no están necesariamente documentados.
     *
     */
    public static function VCI($code)
    {
        $code = strtoupper(strval($code));
        try {

            $codes = [
                'TSY' => 'Autenticación Exitosa.',
                'TSN' => 'Autenticación Rechazada.',
                'NP' => 'No Participa, probablemente por ser una tarjeta extranjera que no participa en el programa 3DSecure, sin autenticación.',
                'U3' => 'Falla conexión, Autenticación Rechazada.',
                'INV' => 'Datos Inválidos.',
                'A' => 'Intentó.',
                'CNP1' => 'Comercio no participa.',
                'EOP' => 'Error operacional.',
                'BNA' => 'BIN no adherido.',
                'ENA' => 'Emisor no adherido.',
                'TO' => 'Tiempo máximo excedido para autenticación.',
                'ABO' => 'Autenticación abortada por tarjeta habiente.',
                'ACS2' => 'Autenticación fallida extranjera.',
                // Para venta extranjera, estos son algunos de los códigos:

                'TSYS' => 'Autenticación exitosa Sin fricción. Resultado autenticación: Autenticación Exitosa.',
                'TSAS' => 'Intento, tarjeta no enrolada / emisor no disponible. Resultado autenticación: Autenticación Exitosa.',
                'TSNS' => 'Fallido, no autenticado, denegado / no permite intentos. Resultado autenticación: Autenticación denegada.',
                'TSRS' => 'Autenticación rechazada - sin fricción. Resultado autenticación: Autenticación rechazada.',
                'TSUS' => 'Autenticación no se pudo realizar por problema técnico u otro motivo. Resultado autenticación: Autenticación fallida.',
                'TSCF' => 'Autenticación con fricción (No aceptada por el comercio). Resultado autenticación: Autenticación incompleta.',
                'TSYF' => 'Autenticación exitosa con fricción. Resultado autenticación: Autenticación exitosa.',
                'TSNF' => 'No autenticado. Transacción denegada con fricción. Resultado autenticación: Autenticación denegada.',
                'TSUF' => 'Autenticación con fricción no se pudo realizar por problema técnico u otro. Resultado autenticación: Autenticación fallida.',
                'NPC' => 'Comercio no Participa. Resultado autenticación: Comercio/BIN no participa.',
                'NPB' => 'BIN no participa. Resultado autenticación: Comercio/BIN no participa.',
                'NPCB' => 'Comercio y BIN no participan. Resultado autenticación: Comercio/BIN no participa.',
                'SPCB' => 'Comercio y BIN sí participan. Resultado autenticación: Autorización incompleta.',

            ];
            return $codes[$code];
        } catch (\Exception $exception) {
            return 'Exception : Código no encontrado.';
        }
    }

    /**
     * ResponseCode
     *
     * @param string $code response_code.
     * @return string retorna el significado del código, Código de respuesta de la autorización.
     *
     */
    public static function ResponseCode($code)
    {
        $code = strval($code);
        try {
            $codes = [
                '0' => 'Transacción aprobada.',
                '-1' => 'Rechazo - Posible error en el ingreso de datos de la transacción.',
                '-2' => 'Rechazo - Se produjo fallo al procesar la transacción, este mensaje de rechazo se encuentra relacionado.',
                '-3' => 'Rechazo - Interno Transbank.',
                '-4' => 'Rechazo - Rechazada por parte del emisor.',
                '-5' => 'Rechazo - Transacción con riesgo de posible fraude.',
            ];
            return $codes[$code];
        } catch (\Exception $exception) {
            return 'Exception : Código no encontrado.';
        }
    }

    /**
     * ResponseCommerceCode
     *
     * @param string $code response_commerce_code.
     * @return string retorna el significado del código, Código de respuesta de la transacción al comercio.
     *
     */
    public static function ResponseCommerceCode($code)
    {
        $code = strval($code);
        try {
            $codes = [
                '-1' => 'Tarjeta inválida.',
                '-2' => 'Error de conexión.',
                '-3' => 'Excede monto máximo.',
                '-4' => 'Fecha de expiración inválida.',
                '-5' => 'Problema en autenticación.',
                '-6' => 'Rechazo general.',
                '-7' => 'Tarjeta bloqueada.',
                '-8' => 'Tarjeta vencida.',
                '-9' => 'Transacción no soportada.',
                '-10' => 'Problema en la transacción.',
                '-11' => 'Excede límite de reintentos de rechazos (Próximamente).',
            ];
            return $codes[$code];
        } catch (\Exception $exception) {
            return 'Exception : Código no encontrado.';
        }
    }

    /**
     * Status
     *
     * @param string $code status.
     * @return string retorna el significado del código, Estado de la transacción, Largo máximo: 64
     *
     */
    public static function Status($code)
    {
        $code = strtoupper(strval($code));
        try {
            $codes = [
                'INITIALIZED' => 'Transacción inicializada.',
                'AUTHORIZED' => 'Transacción autorizada.',
                'REVERSED' => 'Transacción revertida.',
                'FAILED' => 'Transacción fallida.',
                'NULLIFIED' => 'Transacción anulada.',
                'PARTIALLY_NULLIFIED' => 'Transacción parcialmente anulada.',
                'CAPTURED' => 'Transacción capturada.',
            ];
            return $codes[$code];
        } catch (\Exception $exception) {
            return 'Exception : Código no encontrado.';
        }
    }

    /**
     * PaymentTypeCode
     *
     * @param string $code payment_type_code.
     * @return string retorna el significado del código, Tipo de pago de la transacción.
     *
     */
    public static function PaymentTypeCode($code)
    {
        $code = strtoupper(strval($code));
        try {
            $codes = [
                'VD' => 'Venta Débito.',
                'VN' => 'Venta Normal.',
                'VC' => 'Venta en cuotas.',
                'SI' => '3 cuotas sin interés.',
                'S2' => '2 cuotas sin interés.',
                'NC' => 'N Cuotas sin interés.',
                'VP' => 'Venta Prepago.',
            ];
            return $codes[$code];
        } catch (\Exception $exception) {
            return 'Exception : Código no encontrado.';
        }
    }

    /**
     * RefundCode
     *
     * @param string $code refund_code.
     * @return string retorna el significado del código, Reembolso de la transacción.
     *
     */
    public static function RefundCode($code)
    {
        $code = strtoupper(strval($code));
        try {
            $codes = [
                '304' => 'Validación de campos de entrada nulos.',
                '245' => 'Código de comercio no existe.',
                '22' => 'El comercio no se encuentra activo.',
                '316' => 'El comercio indicado no corresponde al certificado o no es hijo del comercio MALL en caso de transacciones MALL.',
                '308' => 'Operación no permitida.',
                '274' => 'Transacción no encontrada.',
                '16' => 'La transacción no permite anulación.',
                '292' => 'La transacción no está autorizada.',
                '284' => 'Periodo de anulación excedido.',
                '310' => 'Transacción anulada previamente.',
                '311' => 'Monto a anular excede el saldo disponible para anular.',
                '312' => 'Error genérico para anulaciones.',
                '315' => 'Error del autorizador.',
                '53' => 'La transacción no permite anulación parcial de transacciones con cuotas.',
            ];
            return $codes[$code];
        } catch (\Exception $exception) {
            return 'Exception : Código no encontrado.';
        }
    }

}


src/WebpayPlus.php
<?php


namespace Innovaweb\Transbank;

use Exception;
use GuzzleHttp\Exception\GuzzleException;
use Innovaweb\Transbank\Helpers\HelperRedirect;
use Transbank\Webpay\WebpayPlus as WPPlus;
use Transbank\Webpay\WebpayPlus\Exceptions\TransactionCreateException;
use Transbank\Webpay\WebpayPlus\Transaction as Transaction;

class WebpayPlus
{
    /** @const INTEGRATION Development environment */
    const INTEGRATION = 'integration';

    /** @const PRODUCTION Production environment */
    const PRODUCTION = 'production';

    public $token;
    public $url;


    /**
     * WebpayPlus constructor
     *
     * @param string $webpay_plus_commerce_code Código de comercio del cliente para WebpayPlus
     * @param string $webpay_plus_api_key Api Key del cliente para WebpayPlus
     * @param string $environment Ambiente de [ 'integration' , 'production']
     */
    public function __construct(string $webpay_plus_commerce_code = '',
                                string $webpay_plus_api_key = '',
                                string $environment = self::INTEGRATION)
    {
        if ($environment === self::PRODUCTION and !empty($webpay_plus_commerce_code) and !empty($webpay_plus_api_key)) {
            WPPlus::configureForProduction($webpay_plus_commerce_code, $webpay_plus_api_key);
        } else {
            WPPlus::configureForTesting();
        }
    }

    /**
     * createTransaction
     *
     * @param string $buy_order Orden de compra de la tienda. Este número debe ser único para cada transacción. Largo máximo: 26. La orden de compra puede tener: Números, letras, mayúsculas y minúsculas, y los signos "|_=&%.,~:/?[+!@()>-"
     * @param string $session_id Identificador de sesión, uso interno de comercio, este valor es devuelto al final de la transacción. Largo máximo: 61
     * @param float $amount Monto de la transacción. Máximo 2 decimales para USD. Largo máximo: 17
     * @param string $url_return URL del comercio, a la cual Webpay redireccionará posterior al proceso de autorización. Largo máximo: 256
     * @return array en caso de success retorna un objeto con token: con Token de la transacción. Largo: 64, string url: URL de formulario de pago Webpay. Largo máximo: 255.
     *
     * @throws GuzzleException
     */
    public function createTransaction($buy_order, $session_id, $amount, $url_return)
    {

        try {

            $response = (new Transaction)->create($buy_order, $session_id, $amount, $url_return);

            $this->token = $response->getToken();
            $this->url = $response->getUrl();

            return [
                'status' => 'success',
                'response' => $response
            ];

        } catch (TransactionCreateException $exception) {
            return [
                'status' => 'error',
                'exception' => $exception->getMessage(),
            ];
        } catch (Exception $exception) {
            return [
                'status' => 'error',
                'exception' => $exception->getMessage(),
            ];
        }
    }

    /**
     * commitTransaction
     *
     * @param string $token_ws Token de la transacción. Largo: 64.
     * @return array en caso de success retorna objecto con datos de la transacción
     *
     * @throws GuzzleException
     */
    public function commitTransaction($token_ws)
    {
        try {

            if (!$token_ws) {
                $token_ws = $_POST['token_ws'];
            }
            $response = (new Transaction)->commit($token_ws);

            return [
                'status' => 'success',
                'response' => $response
            ];

        } catch (Exception $exception) {
            return [
                'status' => 'error',
                'exception' => $exception->getMessage(),
            ];
        }
    }

    /**
     * refundTransaction
     *
     * @param string $token Token de la transacción. Largo: 64.
     * @param float $amount Monto de la transacción. Máximo 2 decimales para USD. Largo máximo: 17
     * @return array en caso de success retorna objecto con datos del reembolso
     *
     * @throws GuzzleException
     */
    public function refundTransaction($token, $amount)
    {
        try {
            $response = (new Transaction)->refund($token, $amount);
            return [
                'status' => 'success',
                'response' => $response
            ];

        } catch (Exception $exception) {
            return [
                'status' => 'error',
                'exception' => $exception->getMessage(),
            ];
        }
    }

    /**
     * getStatusTransaction
     *
     * @param string $token Token de la transacción. Largo: 64.
     * @return array en caso de success retorna objecto con datos de la transacción
     * @throws GuzzleException
     */
    public function getStatusTransaction($token)
    {
        try {
            $response = (new Transaction)->status($token);
            return [
                'status' => 'success',
                'response' => $response
            ];

        } catch (Exception $exception) {
            return [
                'status' => 'error',
                'exception' => $exception->getMessage(),
            ];
        }
    }

    /**
     * getTransactionStatus
     *
     * @param string $token Token de la transacción. Largo: 64.
     * @return array en caso de success retorna objecto con datos de la transacción
     * @throws GuzzleException
     */
    public function getTransactionStatus(string $token)
    {
        return $this->getStatusTransaction($token);
    }

    public function redirectHTML($url = '', $token = '')
    {
        $url = empty($url) ? $this->url : $url;
        $token = empty($token) ? $this->token : $token;
        return HelperRedirect::redirectHTML($url, $token);
    }

    public function dataArray($url = '', $token = '')
    {
        $url = empty($url) ? $this->url : $url;
        $token = empty($token) ? $this->token : $token;
        return HelperRedirect::dataArray($url, $token);
    }
}
